cpp/src/arrow/array/array_binary_test.cc

```diff
@@ -1007,11 +1013,97 @@ class TestBaseBinaryDataVisitor : public ::testing::Test {
 };
-
 TYPED_TEST_SUITE(TestBaseBinaryDataVisitor, BaseBinaryOrBinaryViewLikeArrowTypes);
 
 TYPED_TEST(TestBaseBinaryDataVisitor, Basics) { this->TestBasics(); }
 
 TYPED_TEST(TestBaseBinaryDataVisitor, Sliced) { this->TestSliced(); }
 
+template <typename Type>
+class TestCompactArray : public ::testing::Test {
+  using ArrayType = typename TypeTraits<Type>::ArrayType;
+
+ public:
+  void SetUp() override { type_ = TypeTraits<Type>::type_singleton(); }
+  void TestType() {
+    auto array = internal::checked_pointer_cast<ArrayType>(
+        ArrayFromJSON(type_, R"(["a","b","c"])"));
+    ASSERT_OK_AND_ASSIGN(auto compacted_array, array->CompactArray());
+    ASSERT_EQ(compacted_array->type(), type_);
+  }
+
+  void TestBuffers() {
+    // Check Empty state
+    ASSERT_OK_AND_ASSIGN(auto array, MakeBinaryViewArray<Type>({}, {}));
+    ASSERT_OK_AND_ASSIGN(auto compacted_array, array->CompactArray());
+    ASSERT_EQ(compacted_array->data()->buffers.size(), 2);
+    AssertArraysEqual(*array, *compacted_array);
+
+    auto buffer_a = Buffer::FromString(std::string(1 << 5, 'a'));
+    auto buffer_b = Buffer::FromString(std::string(1 << 5, 'b'));
+
+    // Check the state of no elements in View Buffer
+    ASSERT_OK_AND_ASSIGN(array, MakeBinaryViewArray<Type>({buffer_a, buffer_b}, {}));
+    ASSERT_OK_AND_ASSIGN(compacted_array, array->CompactArray());
+    ASSERT_EQ(compacted_array->data()->buffers.size(), 2);
+    AssertArraysEqual(*array, *compacted_array);
+
+    // Check the state of no reference to data buffers from view elements
+    ASSERT_OK_AND_ASSIGN(
+        array, MakeBinaryViewArray<Type>({buffer_a, buffer_b},
+                                         {
+                                             util::ToInlineBinaryView("inlined 1"),
+                                             util::ToInlineBinaryView("inlined 2"),
+                                         }))
+    ASSERT_OK_AND_ASSIGN(compacted_array, array->CompactArray());
+    ASSERT_EQ(compacted_array->data()->buffers.size(), 2);
+    AssertArraysEqual(*array, *compacted_array);
+
+    // Check the state of there is one reference to whole data of one data buffer
+    ASSERT_OK_AND_ASSIGN(
+        array,
+        MakeBinaryViewArray<Type>(
+            {buffer_a, buffer_b},
+            {
+                util::ToBinaryView("bbbb", static_cast<int32_t>(buffer_b->size()), 1, 0),
+                util::ToInlineBinaryView("inlined 1"),
+            }));
+    ASSERT_OK_AND_ASSIGN(compacted_array, array->CompactArray());
+    ASSERT_EQ(compacted_array->data()->buffers.size(), 3);
+    AssertArraysEqual(*array, *compacted_array);
+
+    // Check the state of there is one reference to partial data of one data buffer
+    ASSERT_OK_AND_ASSIGN(
+        array, MakeBinaryViewArray<Type>(
+                   {buffer_a, buffer_b},
+                   {
+                       util::ToBinaryView(
+                           "bbbb", static_cast<int32_t>(buffer_b->size() >> 1), 1, 16),
```

Comment:
`static_cast<int32_t>(buffer_b->size() >> 1)` is pointlessly cryptic, how about 16 or something?